server/http4s-server/src/main/scala/sttp/tapir/server/http4s/Http4sWebSockets.scala

```diff
@@ -1,114 +1,137 @@
 package sttp.tapir.server.http4s
 
-import cats.Monad
 import cats.effect.Temporal
-import cats.effect.std.Queue
+import cats.{Applicative, Monad}
 import cats.syntax.all._
 import fs2._
+import fs2.concurrent.Channel
 import org.http4s.websocket.{WebSocketFrame => Http4sWebSocketFrame}
 import scodec.bits.ByteVector
 import sttp.capabilities.fs2.Fs2Streams
 import sttp.tapir.model.WebSocketFrameDecodeFailure
 import sttp.tapir.{DecodeResult, WebSocketBodyOutput}
 import sttp.ws.WebSocketFrame
+import cats.effect.implicits.parallelForGenSpawn
 
 private[http4s] object Http4sWebSockets {
   def pipeToBody[F[_]: Temporal, REQ, RESP](
-      pipe: Pipe[F, REQ, RESP],
-      o: WebSocketBodyOutput[Pipe[F, REQ, RESP], REQ, RESP, _, Fs2Streams[F]]
-  ): F[Pipe[F, Http4sWebSocketFrame, Http4sWebSocketFrame]] = {
-    Queue.bounded[F, WebSocketFrame](2).map { pongs => (in: Stream[F, Http4sWebSocketFrame]) =>
-      val sttpFrames = in.map(http4sFrameToFrame)
-      val concatenated = optionallyConcatenateFrames(sttpFrames, o.concatenateFragmentedFrames)
-      val ignorePongs = optionallyIgnorePong(concatenated, o.ignorePong)
-      val autoPongs = optionallyAutoPong(ignorePongs, pongs, o.autoPongOnPing)
-      val autoPings = o.autoPing match {
-        case Some((interval, frame)) => Stream.awakeEvery[F](interval).map(_ => frame)
-        case None                    => Stream.empty
-      }
-      val decodeClose = optionallyDecodeClose(autoPongs, o.decodeCloseRequests)
+                                   pipe: Pipe[F, REQ, RESP],
+                                   o: WebSocketBodyOutput[Pipe[F, REQ, RESP], REQ, RESP, _, Fs2Streams[F]]
+                                 ): F[Pipe[F, Http4sWebSocketFrame, Http4sWebSocketFrame]] = {
+    if ((!o.concatenateFragmentedFrames) && (!o.ignorePong) && (!o.autoPongOnPing) && o.autoPing.isEmpty) {
```

Comment:
I'm not quite sure what's the real difference between the fast & slow tracks - if the flags above are set to appropriate values, apart from some additional initialisation code, the fs2 pipeline seems to end up the same in both cases?

Reply:
Second question: why these particular flag combination has been chosen for the fast track? The [default](https://github.com/softwaremill/tapir/blob/7a9e8872a7e3faaf2f7d94bb7e4ab01beb72c80b/core/src/main/scala/sttp/tapir/Tapir.scala#L176-L190C5) is to have auto pings (which I think is reasonable), auto pongs, ignore pongs

Reply:
The difference between the fast and slow track is that a fast track merely pipes the input stream through the business logic, whereas "slow" track involves a heavier machinery of concurrently merging a) business logic output b) autoPing and c) autoPongOnPing streams. This also answers the second question - if we want to avoid concurrent merging of streams for performance reasons, a certain flag combination is enforced, unfortunately. That said, looking at the flags in the fast track:
- `concatenateFragmentedFrames` is currently a no-op,
- `ignorePong` and `autoPongOnPing` can be handled on the `http4s` level. (`autoPongOnPing` is a bit of an unfortunate flag as it conflates two distinct behaviors - auto pong AND filtering ping. We cosidered adding `WebSocketBodyOutput.ignorePing` but ditched the idea due to binary compatibility issues. )
- admittedly, the one remaing is `autoPing == false`, which has to be handled somewhere else (server layer or busines layer)

Also, even the "slow" track is quite performant compared to the status quo.

Reply:
Thanks for the reply - so in the "slow path" we are allocating a `Channel` to do the concurrent merging.

We are using the channel for two purposes: auto-pong and auto-ping. The other flags simply modify the input / output streams, leaving them unchanged if they are `false`. So maybe we could relax the condition for the fast path, so that it's taken when auto-pong=false & auto-ping=false? I think we should be able to do the concatenating / close-decoding transformations on the fast path as well?

Secondly, I think it would be great to add some documentation, explaining how to best use http4s websockets. As I understand, http4s has also a built-in auto-ping mechanism, which should preferrably be used - that should end up in the docs (a similar note is present in the akka docs, btw.). Similarly, information about the slow/fast paths, what is the performance difference would be great to have, so that people could make an informed choice.

Reply:
Will revise, cannot promise before Christmas, though :)

Reply:
Regrettably, `http4s` lacks the auto-ping mechanism and doing it on the Stream (tapir) level turns out to be a (unbelievably) costly operation. Here is why: merging a stream with an empty(!) stream reduces the speed by `x800`:
```scala
val n = 10_000
val m = 100
val s = Stream(1).covary[IO].repeatN(n)
val s2 = s.mergeHaltL(Stream.empty)

s.compile.last.replicateA_(m).timed.map(_._1.toMillis).flatMap(IO.println) >>
s2.compile.last.replicateA_(m).timed.map(_._1.toMillis).flatMap(IO.println)
```
```
130
105394
```

Reply:
> Regrettably, http4s lacks the auto-ping mechanism and doing it on the Stream (tapir) level turns out to be a (unbelievably) costly operation. Here is why: merging a stream with an empty(!) stream reduces the speed by x800:

That looks ... really bad. To be honest I did not suspect that usage of fs2 might dominate I/O, but I guess that's what we're looking at here. 

Unless we're using fs2 in a totally wrong way ...